### PR TITLE
Remove some Expr clones in `EliminateCrossJoin`(3%-5% faster planning)

### PR DESCRIPTION
**Pull request by alamb**
~Draft as it builds on https://github.com/apache/datafusion/pull/10427~

## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/10287

## Rationale for this change

I am trying to avoid copying as much in `EliminateCrossJoin` to speed up planning. 

One thing this rule does is quite a few `Expr` copies while checking exprs for join keys

## What changes are included in this PR?

1. Move management of join key exprs to `JoinExprSet` struct
2. Reduce copying of clones internally to `JoinExprSet`
3. New tests for `JoinExprSet`

## Are these changes tested?
Existing CI 

## Are there any user-facing changes?
No functional changes

Performance tests show 3-5% faster for TPCH / TPDS queries

<details><summary>Details</summary>
<p>

```
++ critcmp main less_clone_in_eliminate_cross_join
group                                         less_clone_in_eliminate_cross_join     main
-----                                         ----------------------------------     ----
logical_aggregate_with_join                   1.00  1184.8±12.12µs        ? ?/sec    1.02  1207.4±14.82µs        ? ?/sec
logical_plan_tpcds_all                        1.00    156.8±1.94ms        ? ?/sec    1.01    158.1±2.46ms        ? ?/sec
logical_plan_tpch_all                         1.00     16.8±0.24ms        ? ?/sec    1.01     16.9±0.15ms        ? ?/sec
logical_select_all_from_1000                  1.00     17.9±0.11ms        ? ?/sec    1.04     18.7±0.13ms        ? ?/sec
logical_select_one_from_700                   1.00   815.1±20.51µs        ? ?/sec    1.00   813.6±10.21µs        ? ?/sec
logical_trivial_join_high_numbered_columns    1.00   752.3±14.99µs        ? ?/sec    1.02   764.2±14.21µs        ? ?/sec
logical_trivial_join_low_numbered_columns     1.00   736.0±10.70µs        ? ?/sec    1.01    745.7±7.27µs        ? ?/sec
physical_plan_tpcds_all                       1.00   1308.6±9.71ms        ? ?/sec    1.03   1349.1±7.72ms        ? ?/sec
physical_plan_tpch_all                        1.00     88.6±1.59ms        ? ?/sec    1.05     93.0±1.50ms        ? ?/sec
physical_plan_tpch_q1                         1.00      4.7±0.07ms        ? ?/sec    1.08      5.1±0.07ms        ? ?/sec
physical_plan_tpch_q10                        1.00      4.3±0.08ms        ? ?/sec    1.03      4.4±0.09ms        ? ?/sec
physical_plan_tpch_q11                        1.00      3.8±0.08ms        ? ?/sec    1.05      3.9±0.07ms        ? ?/sec
physical_plan_tpch_q12                        1.00      2.9±0.05ms        ? ?/sec    1.11      3.2±0.05ms        ? ?/sec
physical_plan_tpch_q13                        1.00      2.1±0.04ms        ? ?/sec    1.02      2.1±0.03ms        ? ?/sec
physical_plan_tpch_q14                        1.00      2.5±0.06ms        ? ?/sec    1.10      2.8±0.05ms        ? ?/sec
physical_plan_tpch_q16                        1.00      3.6±0.06ms        ? ?/sec    1.07      3.8±0.07ms        ? ?/sec
physical_plan_tpch_q17                        1.00      3.5±0.12ms        ? ?/sec    1.05      3.7±0.08ms        ? ?/sec
physical_plan_tpch_q18                        1.00      3.9±0.07ms        ? ?/sec    1.04      4.1±0.05ms        ? ?/sec
physical_plan_tpch_q19                        1.00      5.8±0.07ms        ? ?/sec    1.08      6.3±0.08ms        ? ?/sec
physical_plan_tpch_q2                         1.00      7.6±0.08ms        ? ?/sec    1.03      7.8±0.11ms        ? ?/sec
physical_plan_tpch_q20                        1.00      4.5±0.10ms        ? ?/sec    1.03      4.6±0.07ms        ? ?/sec
physical_plan_tpch_q21                        1.00      6.1±0.07ms        ? ?/sec    1.02      6.2±0.08ms        ? ?/sec
physical_plan_tpch_q22                        1.00      3.3±0.06ms        ? ?/sec    1.05      3.4±0.07ms        ? ?/sec
physical_plan_tpch_q3                         1.00      3.1±0.06ms        ? ?/sec    1.02      3.1±0.06ms        ? ?/sec
physical_plan_tpch_q4                         1.00      2.2±0.04ms        ? ?/sec    1.04      2.3±0.05ms        ? ?/sec
physical_plan_tpch_q5                         1.00      4.4±0.07ms        ? ?/sec    1.03      4.5±0.09ms        ? ?/sec
physical_plan_tpch_q6                         1.00  1479.7±35.88µs        ? ?/sec    1.09  1616.7±64.90µs        ? ?/sec
physical_plan_tpch_q7                         1.00      5.5±0.11ms        ? ?/sec    1.03      5.7±0.08ms        ? ?/sec
physical_plan_tpch_q8                         1.00      7.2±0.10ms        ? ?/sec    1.02      7.3±0.09ms        ? ?/sec
physical_plan_tpch_q9                         1.00      5.5±0.09ms        ? ?/sec    1.01      5.5±0.07ms        ? ?/sec
physical_select_all_from_1000                 1.00     58.7±0.33ms        ? ?/sec    1.04     61.1±0.87ms        ? ?/sec
physical_select_one_from_700                  1.00      3.6±0.03ms        ? ?/sec    1.02      3.6±0.05ms        ? ?/sec
```

</p>
</details>

### CODE REVIEW

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 1]
```diff
@@ -250,90 +265,67 @@ fn find_inner_join(
     }))
 }
 
-fn intersect(
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:253]: This was moved into JoinExprSet

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 2]
```diff
@@ -250,90 +265,67 @@ fn find_inner_join(
...
         match op {
             Operator::Eq => {
-                // Ensure that we don't add the same Join keys multiple times
-                if !(accum.contains(&(*left.clone(), *right.clone()))
-                    || accum.contains(&(*right.clone(), *left.clone())))
-                {
-                    accum.push((*left.clone(), *right.clone()));
-                }
+                // insert handles ensuring  we don't add the same Join keys multiple times
+                join_keys.insert(left, right);
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:274]: JoinExprSet does not require cloning to check if the exprs should be inserted

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 3]
```diff
@@ -250,90 +265,67 @@ fn find_inner_join(
...
+/// # Returns
+/// * `Some()` when there are few remaining predicates in filter_expr
+/// * `None` otherwise
+fn remove_join_expressions(expr: Expr, join_keys: &JoinKeySet) -> Option<Expr> {
     match expr {
-        Expr::BinaryExpr(BinaryExpr { left, op, right }) => {
-            match op {
-                Operator::Eq => {
-                    if join_keys.contains(&(*left.clone(), *right.clone()))
-                        || join_keys.contains(&(*right.clone(), *left.clone()))
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:313]: there were several `clone`s (deep copies) here too to simply check if the the join keys contained the expressions

File: datafusion/optimizer/src/join_key_set.rs [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+///
+/// This struct models `a.x + 5 = b.y AND a.z = b.z` as two join keys
+/// 1. `(a.x + 5,  b.y)`
+/// 2. `(a.z,      b.z)`
+///
+/// # Important properties:
+///
+/// 1. Retains insert order
+/// 2. Can quickly look up if a pair of expressions are in the set.
+#[derive(Debug)]
```
Comment (alamb) [inline datafusion/optimizer/src/join_key_set.rs:36]: I extracted all the relevant behavior for managing join keys into this file and documented it

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 5]
```diff
@@ -250,90 +265,67 @@ fn find_inner_join(
...
-/// Returns None otherwise
-fn remove_join_expressions(
-    expr: &Expr,
-    join_keys: &HashSet<(Expr, Expr)>,
-) -> Result<Option<Expr>> {
+///
+/// # Returns
+/// * `Some()` when there are few remaining predicates in filter_expr
+/// * `None` otherwise
+fn remove_join_expressions(expr: Expr, join_keys: &JoinKeySet) -> Option<Expr> {
```
Comment (alamb) [inline datafusion/optimizer/src/eliminate_cross_join.rs:300]: This API previously cloned on all paths (including to check join keys) -- the new API does not

Also it never returns `Err` so I changed the signature to `Option` from `Result<Option<..>>`

File: datafusion/optimizer/src/eliminate_cross_join.rs [hunk 6]
```diff
@@ -204,8 +204,8 @@ fn try_flatten_join_inputs(
 fn find_inner_join(
     left_input: &LogicalPlan,
     rights: &mut Vec<LogicalPlan>,
-    possible_join_keys: &[(Expr, Expr)],
-    all_join_keys: &mut HashSet<(Expr, Expr)>,
+    possible_join_keys: &JoinKeySet,
+    all_join_keys: &mut JoinKeySet,
```
Comment (comphead) [inline datafusion/optimizer/src/eliminate_cross_join.rs:208]: that is nice unification

File: datafusion/optimizer/src/join_key_set.rs [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+impl JoinKeySet {
+    /// Create a new empty set
+    pub fn new() -> Self {
+        Self {
+            inner: IndexSet::new(),
+        }
+    }
+
+    /// Return true of this set contains (left = right)
```
Comment (comphead) [inline datafusion/optimizer/src/join_key_set.rs:49]:
```suggestion
    /// Return true if the set contains a join pair 
    /// where left = right or right = left
```

File: datafusion/optimizer/src/join_key_set.rs [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+    /// or (right = left)
```
Comment (comphead) [inline datafusion/optimizer/src/join_key_set.rs:50]:
```suggestion
```

File: datafusion/optimizer/src/join_key_set.rs [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+    pub fn contains(&self, left: &Expr, right: &Expr) -> bool {
+        self.inner.contains(&ExprPair::new(left, right))
+            || self.inner.contains(&ExprPair::new(right, left))
+    }
+
+    /// Insert the join key `(left = right)` into the set  it or `(right =
```
Comment (comphead) [inline datafusion/optimizer/src/join_key_set.rs:56]:
```suggestion
    /// Insert the join key `(left = right)` into the set  if join pair `(right =
```

File: datafusion/optimizer/src/join_key_set.rs [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+        } else {
+            self.inner.insert((left.clone(), right.clone()));
+            true
+        }
+    }
+
+    /// Inserts potentially many join keys into the set, copying only when necessary
+    ///
+    /// returns true if any of the pairs were inserted
+    pub fn insert_many<'a>(
```
Comment (comphead) [inline datafusion/optimizer/src/join_key_set.rs:72]:
```suggestion
    pub fn insert_all<'a>(
```
?

File: datafusion/optimizer/src/join_key_set.rs [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+
+    /// Return the length of this set
+    #[cfg(test)]
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Return an iterator over the join keys in this set
+    pub fn iter(&self) -> impl Iterator<Item = (&Expr, &Expr)> {
+        self.inner.iter().map(|(l, r)| (l, r))
```
Comment (comphead) [inline datafusion/optimizer/src/join_key_set.rs:107]: do we need this map here?

**Review by comphead — COMMENTED** [review]
that is 💪

File: datafusion/optimizer/src/join_key_set.rs (hunk 11 above)
Comment (alamb) [reply datafusion/optimizer/src/join_key_set.rs:107]: Yes: It is (annoying) subtle -- this transforms a iterator of `&(Expr, Expr)` to an iterator of `(&Expr, &Expr)`

It is somewhat annoying that Rust does this conversion implicitly in the map closure but can't do it when in the return type

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

**Comment by alamb** [comment]
Thanks for the review @comphead  🚀

**Pull request merged**